src/libstore/derivations.hh

```diff
@@ -342,16 +346,16 @@ struct Derivation : BasicDerivation
      * 2. Input placeholders are replaced with realized input store
      *    paths.
      */
-    std::optional<BasicDerivation> tryResolve(Store & store) const;
+    std::optional<BasicDerivation> tryResolve(Store & store, Store * evalStore = nullptr) const;
 
     /**
      * Like the above, but instead of querying the Nix database for
-     * realisations, uses a given mapping from input derivation paths +
+     * realisations, uses the given mapping from input derivation paths +
      * output names to actual output store paths.
      */
     std::optional<BasicDerivation> tryResolve(
         Store & store,
-        const std::map<std::pair<StorePath, std::string>, StorePath> & inputDrvOutputs) const;
+        std::function<OutputResolver> outputResolver) const;
```

Comment:
I didn't really like constructing a map in the store case just to store it away. Using function argument to just directly do the lookup we need felt more natural for me. But this can be separate PR --- hydra doesn't need that part.